stdnum/de/steuernummer.py

```diff
@@ -0,0 +1,171 @@
+a lot of actual registered numbers did not abide to the checks in tests.
+>>> compact(' 181/815/0815 5')
+'181/815/08155'
+>>> validate('181/815/08155')
+'181/815/08155'
+validate('201/123/12340', 'Sachsen')
+'201/123/12340'
+>>> validate('4151081508156', 'Thüringen')
+'4151081508156'
+>>> validate('4151181508156', 'Thüringen')
+Traceback (most recent call last):
+    ...
+InvalidFormat: ...
+>>> validate('136695978')
+Traceback (most recent call last):
+    ...
+InvalidLength: ...
+"""
+from stdnum.exceptions import (
+    InvalidLength,
+    InvalidFormat
+)
+from stdnum.util import clean
+
+
+def compact(number):
+    """
+    Convert the number to the minimal representation. This strips the
+    number of any valid separators ' -./,' and removes surrounding whitespace.
+    """
+    number = clean(number, ' -./,').upper().strip()
+    return number
+
+
+state_opening_character = {
+    'Baden-Württemberg': {
+        'standard': '',
+        'bund': '28'
+    },
+    'Bayern': {
+        'standard': '',
+        'bund': '9'
+    },
+    'Berlin': {
+        'standard': '',
+        'bund': '11'
+    },
+    'Brandenburg': {
+        'standard': '0',
+        'bund': '30'
+    },
+    'Bremen': {
+        'standard': '',
+        'bund': '24'
+    },
+    'Hamburg': {
+        'standard': '',
+        'bund': '22'
+    },
+    'Hessen': {
+        'standard': '0',
+        'bund': '26'
+    },
+    'Mecklenburg-Vorpommern': {
+        'standard': '0',
+        'bund': '40'
+    },
+    'Niedersachsen': {
+        'standard': '',
+        'bund': '23'
+    },
+    'Nordrhein-Westfalen': {
+        'standard': '',
+        'bund': '5'
+    },
+    'Rheinland-Pfalz': {
+        'standard': '',
+        'bund': '27'
+    },
+    'Saarland': {
+        'standard': '0',
+        'bund': '10'
+    },
+    'Sachsen': {
+        'standard': '2',
+        'bund': '32'
+    },
+    'Sachsen-Anhalt': {
+        'standard': '1',
+        'bund': '31'
+    },
+    'Schleswig-Holstein': {
+        'standard': '',
+        'bund': '21'
+    },
+    'Thüringen': {
+        'standard': '1',
+        'bund': '41'
+    }
+}
+
+
+def get_state_opening_character(state, schema):
+    """
+    Return a string of the expected openeing chracter of that state.
+    The return value will depend on the specification for that state and the
+    schema type, schema type should be:
+    standard: For old standard schema (Standardschema der Länder) or
+    (Standardschema)
+    bund: For the new federal schema (Vereinheitlichtes Bundesschema) or
+    (Bundesschema)
+    Use state name as the refernce in the wikipedia article.
+    """
+    return state_opening_character[state][schema]
+
+
+def is_number_acceptable_to_state(number, state, schema):
+    """
+    Checks if the number is acceptable for a state.
+    """
+    opening_chars = get_state_opening_character(state, schema)
+    len_opening_chars = len(opening_chars)
+    if number[:len_opening_chars] == opening_chars:
```

Comment:
.startswith() will be more readable. Spent some time to understand what's going on